server/src/main/java/org/opensearch/index/shard/ShardLuceneUtils.java

```diff
@@ -0,0 +1,362 @@
+            } catch (IOException e) {
+                store.markStoreCorrupted(e);
+                throw e;
+            } finally {
+                store.decRef();
+            }
+        }
+    }
+
+    private static void doCheckIndex(Store store, String checkIndexOnStartup, IndexShard indexShard) throws IOException {
+        long timeNS = System.nanoTime();
+        if (!Lucene.indexExists(store.directory())) {
+            return;
+        }
+
+        try (BytesStreamOutput os = new BytesStreamOutput(); PrintStream out = new PrintStream(os, false, StandardCharsets.UTF_8.name())) {
+            if ("checksum".equals(checkIndexOnStartup)) {
+                // physical verification only: verify all checksums for the latest commit
+                IOException corrupt = null;
+                Store.MetadataSnapshot metadata = indexShard.snapshotStoreMetadata();
+                for (Map.Entry<String, StoreFileMetadata> entry : metadata.asMap().entrySet()) {
+                    try {
+                        Store.checkIntegrity(entry.getValue(), store.directory());
+                        out.println("checksum passed: " + entry.getKey());
+                    } catch (IOException exc) {
+                        out.println("checksum failed: " + entry.getKey());
+                        exc.printStackTrace(out);
+                        corrupt = exc;
+                    }
+                }
+                out.flush();
+                if (corrupt != null) {
+                    indexShard.logger.warn("check index [failure]\n{}", os.bytes().utf8ToString());
+                    throw corrupt;
+                }
+            } else {
+                // full checkindex
+                final CheckIndex.Status status = store.checkIndex(out);
+                out.flush();
+                if (!status.clean) {
+                    if (indexShard.state == IndexShardState.CLOSED) {
+                        // ignore if closed....
+                        return;
+                    }
+                    indexShard.logger.warn("check index [failure]\n{}", os.bytes().utf8ToString());
+                    throw new IOException("index check failure");
+                }
+            }
+
+            if (indexShard.logger.isDebugEnabled()) {
+                indexShard.logger.debug("check index [success]\n{}", os.bytes().utf8ToString());
+            }
+        }
+
+        indexShard.recoveryState.getVerifyIndex().checkIndexTime(Math.max(0, TimeValue.nsecToMSec(System.nanoTime() - timeNS)));
+    }
+
+    /**
+     * Downloads segments from given remote segment store for a specific commit.
+     * @param overrideLocal flag to override local segment files with those in remote store
+     * @param sourceRemoteDirectory RemoteSegmentDirectory Instance from which we need to sync segments
+     * @throws IOException if exception occurs while reading segments from remote store
+     */
+    public static void syncSegmentsFromGivenRemoteSegmentStore(
```

Comment:
Does it makes sense to move some methods to RemoteStore directory or remote store specific class?